File: app/src/main/java/org/michaelbel/moviemade/presentation/features/main/appbar/MainAppBar.java
package org.michaelbel.moviemade.presentation.features.main.appbar;

import android.content.Context;
import android.util.AttributeSet;
import android.view.animation.Interpolator;

import com.google.android.material.appbar.AppBarLayout;

import org.michaelbel.moviemade.core.utils.DeviceUtil;

import androidx.annotation.Nullable;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.core.view.ViewCompat;
import androidx.core.view.ViewPropertyAnimatorCompat;
import androidx.interpolator.view.animation.LinearOutSlowInInterpolator;

@CoordinatorLayout.DefaultBehavior(MainAppBarVerticalScrollBehavior.class)
public class MainAppBar extends AppBarLayout {

    private static final Interpolator INTERPOLATOR = new LinearOutSlowInInterpolator();
    private ViewPropertyAnimatorCompat mTranslationAnimator;

    private static final int DEFAULT_ANIMATION_DURATION = 200;

    private int mRippleAnimationDuration = (int) (DEFAULT_ANIMATION_DURATION * 2.5);

    private boolean mAutoHideEnabled = true;
    private boolean mIsHidden = false;

    public MainAppBar(Context context) {
        super(context);
    }

    public MainAppBar(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public void hide() {
        hide(true);
    }

    public void hide(boolean animate) {
        mIsHidden = true;
        setTranslationY(-(this.getHeight() + DeviceUtil.INSTANCE.getStatusBarHeight(getContext())), animate);
    }

    public void show() {
        show(true);
    }

    public void show(boolean animate) {
        mIsHidden = false;
        setTranslationY(0, animate);
    }

    private void setTranslationY(int offset, boolean animate) {
        if (animate) {
            animateOffset(offset);
        } else {
            if (mTranslationAnimator != null) {
                mTranslationAnimator.cancel();
            }
            this.setTranslationY(offset);
        }
    }

    private void animateOffset(final int offset) {
        if (mTranslationAnimator == null) {
            mTranslationAnimator = ViewCompat.animate(this);
            mTranslationAnimator.setDuration(mRippleAnimationDuration);
            mTranslationAnimator.setInterpolator(INTERPOLATOR);
        } else {
            mTranslationAnimator.cancel();
        }
        mTranslationAnimator.translationY(offset).start();
    }

    public boolean isHidden() {
        return mIsHidden;
    }

    public boolean isAutoHideEnabled() {
        return mAutoHideEnabled;
    }
}

File: app/src/main/java/org/michaelbel/moviemade/presentation/common/appbar/MainAppBar.kt
package org.michaelbel.moviemade.presentation.common.appbar

import android.content.Context
import android.util.AttributeSet
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.view.ViewCompat
import androidx.core.view.ViewPropertyAnimatorCompat
import androidx.interpolator.view.animation.LinearOutSlowInInterpolator
import com.google.android.material.appbar.AppBarLayout
import org.michaelbel.moviemade.ktx.statusBarHeight

// fixme use AttachedBehavior
@CoordinatorLayout.DefaultBehavior(MainAppBarVerticalScrollBehavior::class)
class MainAppBar(context: Context, attrs: AttributeSet?): AppBarLayout(context, attrs) {

    private var translationAnimator: ViewPropertyAnimatorCompat? = null
    private val rippleAnimationDuration: Int = (200 * 2.5).toInt()

    val isAutoHideEnabled = true
    var isHidden = false

    fun hide(animate: Boolean) {
        isHidden = true
        setTranslationY(-(this.height + context.statusBarHeight), animate)
    }

    fun show(animate: Boolean) {
        isHidden = false
        setTranslationY(0, animate)
    }

    private fun setTranslationY(offset: Int, animate: Boolean) {
        if (animate) {
            animateOffset(offset)
        } else {
            translationAnimator?.cancel()
            translationY = offset.toFloat()
        }
    }

    private fun animateOffset(offset: Int) {
        if (translationAnimator == null) {
            translationAnimator = ViewCompat.animate(this).apply {
                duration = rippleAnimationDuration.toLong()
                interpolator = LinearOutSlowInInterpolator()
            }
        } else {
            translationAnimator?.cancel()
        }

        translationAnimator?.translationY(offset.toFloat())?.start()
    }
}